x/market/handler/endblock_test.go

```diff
@@ -0,0 +1,213 @@
+	bid6 := createBid(t, originOID, int64(5))
+	bid7 := createBid(t, originOID, int64(8))
+	bid8 := createBid(t, originOID, int64(6))
+	bid9 := createBid(t, originOID, int64(5))
+
+	var winnerTests = []winnerTest{
+		{
+			desc:      "no bids",
+			bids:      []types.Bid{},
+			expWinner: nil,
+			expErr:    errNoBids,
+		},
+		{
+			desc:      "single bid",
+			bids:      []types.Bid{bid9},
+			expWinner: &bid9,
+		},
+		{
+			desc:      "two bids",
+			bids:      []types.Bid{bid7, bid4},
+			expWinner: &bid4,
+		},
+		{
+			desc:      "two matching bids",
+			bids:      []types.Bid{bid9, bid7, bid4},
+			expWinner: &bid4,
+		},
+		{
+			desc:      "all the same bid values",
+			bids:      []types.Bid{bid9, bid9, bid9, bid9, bid9},
+			expWinner: &bid9,
+		},
+		{
+			desc:      "two of the same",
+			bids:      []types.Bid{bid9, bid3},
+			expWinner: &bid3,
+		},
+		{
+			desc:      "multiple of the same, but one lowest value",
+			bids:      []types.Bid{bid1, bid0, bid9, bid2, bid3, bid4},
+			expWinner: &bid0,
+		},
+		{
+			desc:      "confirm last bid in list is picked",
+			bids:      []types.Bid{bid1, bid9, bid2, bid3, bid4, bid0},
+			expWinner: &bid0,
+		},
+		{
+			desc:      "two matching low bids",
+			bids:      []types.Bid{bid5, bid6, bid7, bid8},
+			expWinner: &bid6,
+		},
+	}
+
+	for _, test := range winnerTests {
+		t.Run(test.desc, test.testFunc)
+	}
+}
+
+var errExpUnevenDistribution = errors.New("this test expects uneven winning distribution")
+
+type testDist struct {
+	desc   string
+	bidNum int
+	rounds int
+	expErr error
+}
+
+func (td *testDist) testFunc(t *testing.T) {
+	originOID := testutil.OrderID(t)
+
+	distributionSpread := make(map[int]int)
+	for i := 0; i < td.bidNum; i++ {
+		distributionSpread[i] = 0
+	}
+
+	for i := 0; i < td.rounds; i++ {
+		winCount := make(map[int]int)
+		bIndex := make(map[string]int)
+		bids := make([]types.Bid, 0)
+		// generate N bids all with the same bidding amount
+		for j := 0; j < td.bidNum; j++ {
+			b := createBid(t, originOID, int64(5))
+			bIndex[b.Provider.String()] = j
+			winCount[j] = 0
+			bids = append(bids, b)
+		}
+
+		winner, err := pickBidWinner(bids)
+		if !errors.Is(err, td.expErr) {
+			if !errors.Is(td.expErr, errExpUnevenDistribution) {
+				t.Errorf("returned err: %v does not match %v", err, td.expErr)
+			}
+		}
+		// Check provider
+		slot := bIndex[winner.Provider.String()]
+		distributionSpread[slot] = distributionSpread[slot] + 1
+	}
+
+	for i, v := range distributionSpread {
+		if v == 0 {
+			if td.expErr != nil && !errors.Is(td.expErr, errExpUnevenDistribution) {
+				t.Errorf("index[%d] had no winners", i)
+			}
+		}
+		// calculate a reasonable low expectation of wins given the number of test rounds and number of bidders
+		r := float64(td.rounds/td.bidNum) * 0.5
+		if float64(v) < r {
+			if td.expErr != nil && !errors.Is(td.expErr, errExpUnevenDistribution) {
+				t.Errorf("less than %d winnings of [%d]", td.bidNum, i)
+			}
```

Comment:
```go
if float64(v) < r {
  testutil.Assert(t,td.inconclusive,"underperform",td.bidNum,i)
  continue
}
```

Reply:
Removed the unnecessary nesting by utilizing `continue` above.